assembly/integer/u256.ts

```diff
@@ -463,6 +518,147 @@ export class u256 {
+        dividendLow.hi,
+        remainderHigh.lo,
+        remainderHigh.hi
+      ).div128(divisorLow);
+
+      const quotientLow = res[0];
+      const remainderLow = res[1];
+
+      return [
+        new u256(
+          quotientLow.lo,
+          quotientLow.hi,
+          quotientHigh.lo,
+          quotientHigh.hi
+        ),
+        u256.fromU128(remainderLow),
+      ];
+    }
+
+    // Normalize the divisor and dividend.
+    const shiftAmount: i32 = i32(clz(divisor.hi1) + clz(divisor.hi2));
+    
+    const normalizedDivisor: u256 = divisor << shiftAmount;
+
+    const normalizedDividend: u256 = this >> 1;
+    const normalizedDividendLow: u128 = new u128(
+      normalizedDividend.lo1,
+      normalizedDividend.lo2
+    );
+    const normalizedDividendHigh: u128 = new u128(
+      normalizedDividend.hi1,
+      normalizedDividend.hi2
+    );
+
+    let quotientHigh: u128 = longDivision256by128(
+      normalizedDividendHigh,
+      normalizedDividendLow,
+      new u128(normalizedDivisor.hi1, normalizedDivisor.hi2)
+    )[0];
+    quotientHigh = quotientHigh >> (127 - shiftAmount);
+
+    // Force the quotient to be an underestimate.
+    if (quotientHigh != u128.Zero) {
+      quotientHigh -= u128.One;
+    }
+
+    // Calculate the remainder using the quotient.
+    let quotient: u256 = u256.fromU128(quotientHigh);
+    let remainder: u256 = this - divisor * quotient;
+
+    // Adjust the quotient and remainder if necessary.
+    if (remainder >= divisor) {
+      quotient += u256.One;
+      remainder -= divisor;
+    }
+
+    return [quotient, remainder];
+  }
+
+  /**
+   * Divides a 256-bit number by a 128-bit divisor.
+   * 
+   * @param divisor - The 128-bit divisor.
+   * @returns An array containing the quotient and remainder, both as u128.
+   * @throws {RangeError} If the divisor is zero or if the result will be larger than 128 bits.
+   * 
+   * @remarks
+   * The function employs a long division method adapted for 256-bit numbers divided by 128-bit numbers.
+   * The algorithm starts by normalizing the divisor and dividend to align their most significant bits.
+   * This is done by shifting both the divisor and the dividend by the count of leading zeros in the divisor.
+   * The function then calls `longDivision256by128` to perform the actual division on the normalized numbers.
+   * Finally, the remainder is denormalized by shifting it back to the right by the same amount used to normalize the divisor.
+   */
+  div128(divisor: u128): u128[] {
+    const dividendHigh: u128 = new u128(this.hi1, this.hi2);
+    const dividendLow: u128 = new u128(this.lo1, this.lo2);
+
+    if (divisor == u128.Zero) throw new RangeError("Division by zero");
+
+    // Panic if the result won't fit in a 128-bits number.
+    if (dividendHigh >= divisor) throw new RangeError("Integer overflow");
+
+    const shiftAmount: i32 = i32(u128.clz(divisor));
+
+    const normalizedDivisor: u128 = safeShl(divisor, shiftAmount);
```

Comment:
Using safe shifts is redundant due to `clz` result will never exceed more than `127` for `divisor`. `128` also cover by earlier `divisor == u128.Zero` check.

Basically all `safeShl` and `safeShl` doesn't make sense here and just add extra perf and size overhead.

Reply:
You're quiet right, except for here: 
```typescript
const normalizedDividendHigh: u128 = safeShl(dividendHigh, shiftAmount) | safeShr(dividendLow, 128 - shiftAmount);
```
When shiftAmount = 0, the code right shits with a value of 128 which should be 0, but as the implementation silently overflow, the returned result is wrong. Instead of just handling specifically this case here, I choose a safe implementation of right and left shift.

Do you prefer to remove everything else and handle this specific case by hand ?

Reply:
`shiftAmount == 0` could be handled separately. `shiftAmount == 0` means `divisor == u256.Max` which is special case

Reply:
> As a side note shiftAmount == 0 <=> divisor > 2^128, not necessarily divisor == u256.Max

Yes, but in helper routines which uses `u256 / u128` like (div128 and longDivision256by128) divisor has `u128` type